File: DesignPatterns/01_GOF23/DP_examples/Go/03_Behavioral/Memento/memento.go
package Memento

// Memento 备忘录类
type Memento struct {
	states  map[string]interface{}
	Version int
}

// NewMemento 创建备忘录
func NewMemento(version int) *Memento {
	return &Memento{
		states:  make(map[string]interface{}),
		Version: version,
	}
}

// GetState 获取状态
func (m *Memento) GetState(identifier string) interface{} {
	if value, ok := m.states[identifier]; ok {
		return value
	}
	return nil
}

// SetState 设置状态
func (m *Memento) SetState(identifier string, state interface{}) {
	m.states[identifier] = state
}

File: DesignPatterns/01_GOF23/DP_examples/CSharp/03_Behavioral/Memento/Memento.cs
namespace Behavioral.Memento;

public class Memento
{
    Dictionary<string, object> states;
    public int Version { get; private set; }
    public Memento(int version)
    {
        this.Version = version;
        states = new Dictionary<string, object>();
    }
    public object GetState(string identifier)
    {
        if (states.TryGetValue(identifier, out object? value))
            return value;
        return null;
    }
    public void SetState(string identifier, object state)
    {
        states.Add(identifier, state);
    }
}